src/OpenTelemetry.Api/Context/Propagation/TraceContextPropagator.cs

```diff
@@ -0,0 +1,265 @@
+// <copyright file="TraceContextPropagator.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using OpenTelemetry.Internal;
+
+namespace OpenTelemetry.Context.Propagation
+{
+    /// <summary>
+    /// A text map propagator for W3C trace context. See https://w3c.github.io/trace-context/.
+    /// </summary>
+    public class TraceContextPropagator : TextMapPropagator
+    {
+        private const string TraceParent = "traceparent";
+        private const string TraceState = "tracestate";
+
+        private static readonly int VersionPrefixIdLength = "00-".Length;
+        private static readonly int TraceIdLength = "0af7651916cd43dd8448eb211c80319c".Length;
+        private static readonly int VersionAndTraceIdLength = "00-0af7651916cd43dd8448eb211c80319c-".Length;
+        private static readonly int SpanIdLength = "00f067aa0ba902b7".Length;
+        private static readonly int VersionAndTraceIdAndSpanIdLength = "00-0af7651916cd43dd8448eb211c80319c-00f067aa0ba902b7-".Length;
+        private static readonly int OptionsLength = "00".Length;
+        private static readonly int TraceparentLengthV0 = "00-0af7651916cd43dd8448eb211c80319c-00f067aa0ba902b7-00".Length;
```

Comment:
This one probably can be `TraceParent` instead of `Traceparent` (we've finished the long time debate).